Simulator

stepanvalyavskiy opened the pull request:
Closes #173 - new simulator that corresponds to wp's workflow

g4s8 commented on .gitignore:
```diff
@@ -20,3 +20,6 @@
 dist/**/*
 degitx
 degitx-gitaly
+/simulator/target/maven-status/maven-compiler-plugin/compile/default-cli/inputFiles.lst
```
@stepanvalyavskiy I suppose we can ignore everything in `/simulator/target/**` dir, isn't it?

g4s8 commented on simulator/pom.xml:
```diff
@@ -0,0 +1,78 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>wtf-simulator</artifactId>
+    <groupId>wtf</groupId>
```
@stepanvalyavskiy let's use `org.cqfn`

g4s8 commented on simulator/pom.xml:
```diff
@@ -0,0 +1,78 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>wtf-simulator</artifactId>
```
@stepanvalyavskiy and `degitx-simulator`

g4s8 commented on simulator/pom.xml:
```diff
@@ -0,0 +1,78 @@
+            </snapshots>
+            <id>jcenter</id>
+            <name>jcenter</name>
+            <url>https://jcenter.bintray.com/</url>
```
@stepanvalyavskiy the bintray will be closed soon (in May as I remember), do we need this repository?

stepanvalyavskiy replied:
It's temporal, when I will be ready with transfer paxos code to dgitx, we will no more need in bintray.

g4s8 commented on simulator/resource/default_cfg.yml:
```diff
@@ -0,0 +1,31 @@
+# Restart proposing, if decision wasn't made in given time.
```
@stepanvalyavskiy since we're using yaml for config, I'd suggest being more yamly: group related properties(e.g. `frontend` or `backend`), use group names as enable option (e.g. `dropRate`). This config could be refactored into something like this:
```yaml
frontend:
  proposerTimeout: 300
  replicas: 1
  crashRate: 0.1
backend:
  replicas: 3
  crashRate: 0
network:
  dropRate: 0.1
  async: true
  delay: 0.5
```

g4s8 commented on simulator/src/dgitx/backend/AcceptorManager.kt:
```diff
@@ -0,0 +1,58 @@
+import java.util.concurrent.atomic.AtomicReference
+
+//TODO rewrite java code
+class AcceptorManager(
```
@stepanvalyavskiy what is acceptor manager?

g4s8 commented on simulator/src/dgitx/backend/AsyncBackend.kt:
```diff
@@ -0,0 +1,56 @@
+import wtf.g4s8.examples.spaxos.Acceptor
+import wtf.g4s8.examples.spaxos.Proposal
+
+class AsyncBackend(private val origin: BNode, private val cancel: CompletableJob) : BNode {
```
@stepanvalyavskiy no any documentation here too

g4s8 commented on simulator/src/dgitx/backend/Backend.kt:
```diff
@@ -0,0 +1,120 @@
+import wtf.g4s8.examples.spaxos.*
+import java.util.concurrent.ConcurrentHashMap
+
+class Backend(
```
@stepanvalyavskiy does this represent a node instance? please add javadocs to all type definitions

g4s8 reviewed (changes requested):
@stepanvalyavskiy please check my comments

g4s8 commented:
@stepanvalyavskiy can you please resolve review comments and we merge this branch

stepanvalyavskiy commented:
> @stepanvalyavskiy can you please resolve review comments and we merge this branch

OK, working on it.

stepanvalyavskiy commented:
Hi, @g4s8, I added docs for all components and fixed small problems.
I suggest to merge and create separate Tickets for:
 - yamlize config.
 - removing java code and bintray dependencies.

g4s8 reviewed (approved):
@stepanvalyavskiy thanks
